pkg/archive/archive.go

```diff
@@ -856,179 +856,189 @@ func Tar(path string, compression Compression) (io.ReadCloser, error) {
 	return TarWithOptions(path, &TarOptions{Compression: compression})
 }
 
-// TarWithOptions creates an archive from the directory at `path`, only including files whose relative
-// paths are included in `options.IncludeFiles` (if non-nil) or not in `options.ExcludePatterns`.
-func TarWithOptions(srcPath string, options *TarOptions) (io.ReadCloser, error) {
+// TarWithOptionsTo writes `srcPath` as an uncompressed tar archive to
+// `dest`, only including files whose relative paths are included in
+// `options.IncludeFiles` (if non-nil) or not in
+// `options.ExcludePatterns`.
+func TarWithOptionsTo(dest io.WriteCloser, srcPath string, options *TarOptions) error {
 	// Fix the source path to work with long path names. This is a no-op
 	// on platforms other than Windows.
 	srcPath = fixVolumePathPrefix(srcPath)
 
+	// make changes thread-safe.
```

Comment:
“thread-safe” is not really the point; this is equally problematic for an unthreaded caller which doesn’t expect `options` to be modified.

… and in this case, I think just having a local variable for `IncludeFiles` would also work, without making a large copy of the whole `TarOptions`; that copy ~negates all benefits from passing it by reference.